rootex/framework/entity_factory.cpp

```diff
@@ -313,3 +312,87 @@ void EntityFactory::deleteEntity(Ref<Entity> entity)
 	m_Entities.erase(entity->getID());
 	entity.reset();
 }
+
+String EntityFactory::saveEntityAsClass(Ref<Entity> entity)
+{
+	if (OS::IsDirectory("game/assets/classes/" + entity->getName()))
+	{
+		WARN("Class with name \"" + entity->getName() + "\" already exists. New Class not created.");
+		return nullptr;
+	}
+	else
+	{
+		OS::CreateDirectoryName("game/assets/classes/" + entity->getName());
+	}
+	return saveEntityAsClassRecursively(entity, "game/assets/classes/" + entity->getName() + "/");
+}
+
+String EntityFactory::saveEntityAsClassRecursively(Ref<Entity> entity, const String& path)
+{
+	Ref<HierarchyComponent> hierarchyComponent = entity->getComponent<HierarchyComponent>();
+	JSON::json entityJSON = entity->getJSON();
+	if (hierarchyComponent->getChildren().size() > 0)
+	{
+		Vector<String> children;
+		for (EntityID child : hierarchyComponent->m_ChildrenIDs)
+		{
+			children.push_back(saveEntityAsClassRecursively(EntityFactory::GetSingleton()->getEntities().at(child), path));
+		}
+		entityJSON["Components"]["HierarchyComponent"]["children"] = children;
+	}
+	entityJSON["Components"]["HierarchyComponent"]["parent"] = 1;
+	entityJSON["Entity"].erase("ID");
+	InputOutputFileStream file = OS::CreateFileName(path + entity->getName() + ".entity.json");
+	file << entityJSON.dump(4) << std::endl;
+	file.close();
+	return path + entity->getName() + ".entity.json";
+}
+
+Ref<Entity> EntityFactory::createEntityFromClass(const JSON::json& entityJSON)
+{
+	Ref<Entity> createdEntity = createEntityFromClassRecursively(entityJSON);
+	fixParentIDRecursively(createdEntity, 1);
```

Comment:
again, replace with Macro, and wouldn't this run on ALL the entities rather than ALL NEW entities, I guess you can provide the ID of the root node of the entity tree you just instantiated instead.

Reply:
This will run on ALL NEW entities only.